### PR TITLE
Object template rework

### PR DESCRIPTION
**Pull request by DjWarmonger**
Work in progress:

- [x] Move from vector of vectors to boost::multi_array
- [x] Unify interface, clean clean up magic numbers
- [x] Precalculate offsets for better performance
- [ ] Backward save compatibility
- [ ] Unit tests for regular and edge cases

### CODE REVIEW

File: lib/serializer/BinarySerializer.h [hunk 1]
```diff
@@ -344,6 +344,30 @@ class DLL_LINKAGE BinarySerializer : public CSaverBase
...
 	}
+
+	template <typename T>
+	void save(const boost::multi_array<T, 2> & data)
+	{
+		ui32 length = data.num_elements();
+		*this& length;
+		auto shape = data.shape();
+		ui32 x = shape[0], y = shape[1];
+		*this& x& y;
```
Comment (alexvins) [inline lib/serializer/BinarySerializer.h]: Too ugly, use `save` calls
